platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,277 @@
+Pre-Consensus Recording module, when then sends these events to the Gossip module to be sent out. This loop is
+intentional, with the assumption that the additional latency in this loop will be relatively short (on the order of
+microseconds).
+
+#### Peer Discipline
+
+If a peer is misbehaving, the Gossip module will notify the Bad Node module that one of its peers is misbehaving. For
+example, if a peer is not responding to requests, even after repeated attempts to reconnect with it, it may be "bad".
+Or if the peer is sending events that exceed an acceptable rate, or exceed an acceptable side, then it is "bad". Or if
+the events it sends cannot be parsed, then it is "bad". The Gossip module design may define arbitrary additional 
+
+If the Bad Node module decides that the peer should be penalized, then it will instruct the Gossip module to "shun" that
+peer. "Shunning" is a unilateral behavior that one node can take towards another, where it terminates the connection and
+refuses to work further with that node. If the Bad Node module decides to welcome a peer back into the fold, it can
+instruct the Gossip module to "welcome" the peer back.
+
+There are many potential inputs to the Bad Node module, and therefore, the decision to shun or welcome a node is
+**not** made by the Gossip module, but rather, the Gossip module must let the Bad Node module know of peer
+misbehavior, and the Gossip module must be able to shun and welcome peers on demand.
+
+#### Falling Behind
+
+It may be that during gossip, the Gossip module finds that it cannot continue. Perhaps all connections have failed, or
+perhaps the node is so far behind that none of its peers are able to supply any events. In this case, the node is unable
+to operate, and must reconnect. Gossip will make a call through the Consensus module interface to notify Execution that
+gossip is "down". Execution will then initiate reconnect.
+
+Fundamentally, Execution is responsible for reconnect, but it cannot differentiate from a quiescent network or a
+behind node just by looking at the lack of rounds coming from Consensus. So Consensus **must** tell Execution if it
+knows that it cannot proceed in creating rounds because the node itself is bad, vs. a quiescent network which is
+able to create rounds (if there were any transactions).
+
+#### Roster Changes
+
+At runtime, it is possible that the roster will change dynamically (as happens with the dynamic address book feature).
+Roster changes at the gossip level influences which peers the module will work with. Because different peers may be
+at different points in consensus time, it may be that Alice has a newer roster than this node does, or Bob may have an
+older roster. Bob still needs to be able to gossip, even if he is behind. So each node will need to be able to honor
+connections from all peers in all rosters associated with rounds that have not expired.
```

Comment:
> associated with rounds that have not expired

I'm still not convinced this is the way to go. 

At the end of the day, the network doesn't care about nodes that have been removed from the address book. Permitting nodes that aren't part of the network anymore to continue to gossip is increasing our attack surface area AND code complexity, for no perceivable benefit.

Even if we *did* care about the experience of removed nodes, what does the operator of a removed node stand to gain by getting to participate for another couple rounds?